File: TaPE/06.03/src/card.cpp
#include "card.h"
#include "includes.h"
#include "rand_card.h"

bool is_unique(Cell line[BINGO_SIZE], uint cell_index)
{
    if (cell_index == 0)
        return true;
    for (uint i = 0; i < cell_index - 1; i++)
    {
        // clog << "[LOG] Cell number " << line[i].number << "; " << line[cell_index].number << "; cell index " << cell_index << "; index " << i << endl;

        if (line[i].number == line[cell_index].number)
            return false;
    }
    return true;
}

Card create_card()
{
    Card new_card;

    for (uint i = 0; i < BINGO_SIZE; i++)
    {
        // clog << "[LOG] Start Generate line[" << i << "]" << endl;
        for (uint j = 0; j < BINGO_SIZE; j++)
        {
            uint
                min = 1 + (i * 15),
                max = min + 15;
            // clog << "[LOG] Start Generate cell[" << j << "] min " << min << " max " << max << endl;
            new_card.lines[i][j] = {0};
            do
            {
                uint number = generate_number(min, max);
                new_card.lines[i][j] = Cell{number};
            } while (!is_unique(new_card.lines[i], j));
        }
    }
    return new_card;
}

void check_cell(Card &card, uint value)
{
    for (uint i = 0; i < BINGO_SIZE; i++)
    {
        for (uint j = 0; j < BINGO_SIZE; j++)
        {
            if (card.lines[i][j].number == value)
            {
                card.lines[i][j].checked = true;
                return;
            }
        }
    }
}

void out_card(Card &our_card)
{
    const char BINGO[BINGO_SIZE]{'B', 'I', 'N', 'G', 'O'};
    for (int i = 0; i < BINGO_SIZE; i++)
    {
        cout << BINGO[i] << " | ";

        for (uint j = 0; j < BINGO_SIZE; j++)
        {
            Cell x = our_card.lines[i][j];
            /*
            if (x.checked)
            {
                cout << "[" << x.number << "]"
                     << " ";
            }
            else
            {
            }
            */
            cout << x.number << " (" << (x.checked ? "X" : " ") << ") ";
        }
        cout << std::endl;
    }
}

bool is_winner(Card &card)
{
    for (uint i = 0; i < BINGO_SIZE; i++)
    {
        uint checked_count = 0;
        for (uint j = 0; j < BINGO_SIZE; j++)
        {
            if (card.lines[i][j].checked)
                checked_count++;
        }
        if (checked_count == BINGO_SIZE)
            return true;
    }
    return false;
}

File: TaPE/06.03/c#/src/Card.cs
using System.Text;

struct Card
{
    public const uint BingoSize = 5;
    public Card()
    {
        Cells = new Cell[5, 5];
        Random r = new Random();
        for (uint i = 0; i < BingoSize; i++)
        {
            for (uint j = 0; j < BingoSize; j++)
            {
                var start = 1 + i * 15;
                var end = start + 15;
                Cells[i, j] = new(r.Next(start, end));
            }
        }
    }
    public Cell[,] Cells;
    public void CheckCell(uint number)
    {
        for (int i = 0; i < BingoSize; i++)
        {
            for (int j = 0; j < BingoSize; j++)
            {
                if (Cells[i, j].Number == number)
                {
                    Cells[i, j].IsChecked = true;
                }
            }
        }
    }
    public static char[] Bingo = { 'B', 'I', 'N', 'G', 'O' };
    public override string ToString()
    {
        StringBuilder sb = new();

        for (int i = 0; i < BingoSize; i++)
        {
            sb.Append($"{Bingo[i]} ");
            for (int j = 0; j < BingoSize; j++)
            {
                sb.Append($"{Cells[i, j]} ");
            }
            sb.Append("\n");
        }
        return sb.ToString();
    }
    public bool IsWinner()
    {
        for (int i = 0; i < BingoSize; i++)
        {
            for (int j = 0; j < BingoSize; j++)
            {
                if (!Cells[i, j].IsChecked)
                {
                    return false;
                }
            }
        }
        return true;
    }
}
struct Cell
{
    public Cell(uint number) =>
        Number = number;

    public uint Number;
    public bool IsChecked = false;

    public override string ToString()
    {
        if (IsChecked)
        {
            return $"[{Number}]";
        }

        return $" {Number} ";
    }
}
